websites/management/commands/markdown_cleaning/utils.py

```diff
@@ -0,0 +1,207 @@
+            return self.website_contents[(website_id, dirpath, filename)]
+        except KeyError:
+            dirpath = self.standardize_dirpath(site_relative_path)
+            filename = "_index"
+            return self.website_contents[(website_id, dirpath, filename)]
+
+
+class UrlSiteRelativiser:
+    """
+    Given a possibly-legacy, root-relative url returns a tuple (site,
+    course_relative_url).
+
+    Example 1:
+        /courses/architecture/4-601-introduction-to-art-history-fall-2018/assignments/4.601-second-paper
+        becomes:
+        "/assignments/4.601-second-paper"
+
+    Example 2:
+        "/courses/18-02sc-multivariable-calculus-fall-2010"
+        becomes
+        "/"
+
+    Example 3:
+        "/resources/res-21g-01-kana-spring-2010/katakana"
+        becomes
+        "/katakana"
+    """
+
+    def __init__(self):
+        websites = Website.objects.all()
+        self.website_lookup = {w.name: w for w in websites}
+
+    def __call__(self, url: str):
+        parsed = urlparse(url)
+        path = parsed.path
+        pieces = path.split("/")
+        try:
+            site_index, site_name = next(
+                (i, name)
+                for i, name in enumerate(pieces)
+                if name in self.website_lookup
+            )
+        except StopIteration as err:
+            raise ValueError(f"'{url} does not contain a website name.") from err
+        site_relative_path = "/" + "/".join(pieces[site_index + 1 :])
+
+        site_relative_url = site_relative_path
+        if parsed.fragment:
+            site_relative_url += f"#{parsed.fragment}"
+        return self.website_lookup[site_name], site_relative_url
+
+
+class LegacyFileLookup:
+    """
+    Find content by legacy filename.
+
+    Example: In site "21h-104j...", find:
+        MIT21H_104JF10_syllf09.pdf
+    Matches:
+        WebsiteContent(
+            file="/courses/21h-104j-riots-strikes-and-conspiracies-in-american-history-fall-2010/95e03c4c924a62a8e3876d49f51889c0_MIT21H_104JF10_syllf09.pdf",
+            website=Website(name="21h-104j-riots-strikes-and-conspiracies-in-american-history-fall-2010"),
+            # NOT used for matching!
+            filename="mit21h_104jf10_syllf09",
+            dirpath="content/resources",
+        )
+
+    NOTE: The match is based solely on Website and the filename at end of
+    WebsiteContent.file. (The value "95e03c4c924a62a8e3876d49f51889c0_MIT21H_104JF10_syllf09.pdf"
+    above).
+    Advantages:
+        - case sensitive
+        - includes file extension
+    Disadvantages:
+        - sometimes there's not a unique match (see below).
+
+    For example, in res-21g-01-kana-spring-2010, there are three files:
+        /courses/res-21g-01-kana-spring-2010/UUID1_yokudeki.gif
+        /courses/res-21g-01-kana-spring-2010/UUID2_yokudeki.gif
+        /courses/res-21g-01-kana-spring-2010/UUID3_yokudeki.gif
+    with different parent content.
```

Comment:
See https://mitodl.slack.com/archives/GQ89D0Z5M/p1646334160580419

I'll be creating an issue for this today.

> NOTE: The match is based solely on Website and the filename at end of
>  WebsiteContent.file. (The value "95e03c4c924a62a8e3876d49f51889c0_MIT21H_104JF10_syllf09.pdf"
> above).

I'm not super happy with the approach, but I think it's OK when the file match is unique.

Reply:
It looks like your `ocw-to-hugo` PR to add in `parent_id` was merged?  Can we use `overwrite_ocw_course_content` in `ocw-studio` to bring in `parent_id` and then in a future PR make this association?

Reply:
That's what I was thinking.

We could wait until that work to alter ANY file links, but when there's a unique match the filename (as determined by last bit of `WebsiteContent.file`, not `WebsiteContent.filename`) I feel comfortable making the replacement. (And that willl fix a bunch of links.)